Use an EOA for our spender, simplify tutorial (#92)
coinbase/smart-wallet-docs: docs/pages/guides/spend-permissions/quick-start.mdx

Developer Platform API Key, we'll get one of those later. When prompted to use Coinbase Smart Wallet select "yes".

```bash

> bun create onchain@0.0.12

Reviewer: ```suggestion
bun create onchain@latest
```

From here, we'll modify the app to assemble, sign, approve and use a spend permission to spend our users' funds!

### Set up your spender wallet and environment 

Add the following variables to your `.env`:

```
SPENDER_PRIVATE_KEY=
NEXT_PUBLIC_SPENDER_ADDRESS=
NEXT_PUBLIC_CDP_API_KEY=
BASE_SEPOLIA_PAYMASTER_URL=
```

<Callout type="danger">
Always secure your private keys appropriately!  
We insecurely use an environment variable in this demo for simplicity.
</Callout>

> Our spender will need to sign transactions from our app, so we'll create a local wallet for our spender. We'll need a private key and the corresponding ethereum
> address in our environment variables. You can use a development private key you already have, or generate a random new one. 
> If you have [Foundry](https://book.getfoundry.sh/) installed, you can generate a new wallet via `cast wallet new`. Assign the private key and address to
> `SPENDER_PRIVATE_KEY` and `NEXT_PUBLIC_SPENDER_ADDRESS`, respectively.

Reviewer 2: ```suggestion
Our spender will need to sign transactions from our app, so we'll create a wallet (private key and address) for our spender. 
If you have [Foundry](https://book.getfoundry.sh/) installed, you can generate a new wallet via `cast wallet new`. Assign the private key and address to
`SPENDER_PRIVATE_KEY` and `NEXT_PUBLIC_SPENDER_ADDRESS`, respectively.
```

### Create a spender client

Our client is what our app will use to communicate with the blockchain.

Create a sibling directory to `app` called `lib` and add the following `spender.ts` file to create your spender client.

```ts [lib/spender. ts]
import { createPublicClient, createWalletClient, Hex, http } from "viem";
import { baseSepolia } from "viem/chains";
import { privateKeyToAccount } from "viem/accounts";

export async function getPublicClient() {
  const client = createPublicClient({
    chain: baseSepolia,
    transport: http(),
  });
  return client;
}

export async function getSpenderWalletClient() {
  const spenderAccount = privateKeyToAccount(
    process.env.SPENDER_PRIVATE_KEY! as Hex
  );

  const spenderWallet = await createWalletClient({
    account: spenderAccount,
    chain: baseSepolia,
    transport: http(),
  });
  return spenderWallet;
}
```

### Configure the Smart Wallet URL

> In `app/providers.tsx`, update the value of `keysUrl` to be `"https://keys-beta.coinbase.com/connect"`.

Reviewer: from another PR, this is not `keys-dev.coinbase.com` so make sure to rebase or change this!

> In `app/providers.tsx`, update the value of `keysUrl` to be `"https://keys-beta.coinbase.com/connect"`.

Reviewer: ```suggestion
In `app/providers.tsx`, update the value of `keysUrl` to be `"https://keys-dev.coinbase.com/connect"`.
```

This will point your app to connect to our public dev environment for smart wallet connections.

We also want to point our chain id to Base Sepolia testnet by setting replacing all instances of `base`
with `baseSepolia` in this file (including the import). 

Your config in `app/providers.tsx` should now look like this:

```
const config = createConfig({
  chains: [baseSepolia],
  connectors: [
    coinbaseWallet({
      appName: process.env.NEXT_PUBLIC_ONCHAINKIT_PROJECT_NAME,
      preference: process.env.NEXT_PUBLIC_ONCHAINKIT_WALLET_CONFIG as
        | "smartWalletOnly"
        | "all",
      // @ts-ignore

>       keysUrl: "https://keys-beta.coinbase.com/connect"

Reviewer: ```suggestion
      keysUrl: "https://keys-dev.coinbase.com/connect"
```

We also want to point our chain id to Base Sepolia testnet by setting replacing all instances of `base`
with `baseSepolia` in this file (including the import). 

> Your config in `providers.tsx` should now look like this:

Reviewer 2: ```suggestion
Your config in `app/providers.tsx` should now look like this:
```

1. calls `approveWithSignature` to approve the spend permission
2. calls `spend` to make use of our allowance and spend our user's funds

### Try out your app

Run your app locally with `bun run dev` and visit `localhost:3000`.

When you click the "Subscribe" button you should be prompted to create or connect your Smart Wallet.

> You can create a new Smart Wallet via the popup. Note that you'll need a little ETH in this wallet to fund the
> deployment of your account. If you don't have any testnet ETH, try this [Coinbase faucet](https://portal.cdp.coinbase.com/products/faucet).
> 
> Before you continue, note that you'll also need a little testnet ETH in your spender wallet, which will need to fund the gas for the transactions.

Reviewer 2: ```suggestion
Note that we'll need a little bit of base sepolia ETH in both wallet addresses (the "user" wallet and the "app" wallet). In a more involved implementation you would use a paymaster to eliminate this requirement. For now, If you don't have any base sepolia ETH, try this [Coinbase faucet](https://portal.cdp.coinbase.com/products/faucet).
```

Do we actually require funds in the user wallet? Wouldn't the dapp's wallet be deploying the user's account during execution?

Author: iirc I've actually seen the popup prompt me to deploy the wallet.... though it's been a while since I remember that happening.  I just tried this with a fresh smart wallet and before I had any ETH i see this screen: 
![Uploading Screenshot 2024-11-07 at 6.57.36 PM.png…]()

Author: once I funded the wallet, I was not prompted to deploy the wallet by the popup, we moved right on to spendpermissions

Author: while I don't ever see the wallet prompt me to deploy it, I do see a warning screen if I have 0 ETH in the wallet with some buttons including the option to "Fund wallet".  So while the user may not require ETH in a functional sense, something in our system is enforcing that

Author: (ultimately they will require at least a little ETH if the dev wants to see the app work end-to-end because there needs to be a little for the spender to spend)

Reviewer 2: Gotcha. If the dapp requests a base asset (i.e. ETH) then the user will be prompted to fund before granting (so the dapp has something to spend). If the dapp requests an erc20 in an ideal world we'd probably ask the user to fund there as well, but in the current world we just let the user grant the permission without funding.

Once your wallet is created and both wallets are funded, return to the app and click "Subscribe".

> Sign the prompts, which will deploy your new Smart Wallet (if undeployed), and then prompt you to sign the spend permission requested by the app.

Reviewer 2: I don't think we prompt to deploy here - and if we did for base sepolia the default paymaster would handle funds for us

Author: got it -- yeah as I mentioned in my comment above, I don't see this when I just tried it fresh so will remove
